rbs/TypeTranslator.cc

```diff
@@ -0,0 +1,416 @@
+                    e.setHeader("Unexpected node type `{}` in record key type, expected `{}`",
+                                rbs_node_type_name(hash_node->key), "Symbol");
+                }
+                continue;
+            }
+        }
+
+        if (hash_node->value->type != RBS_TYPES_RECORD_FIELDTYPE) {
+            if (auto e = ctx.beginError(loc, core::errors::Internal::InternalError)) {
+                e.setHeader("Unexpected node type `{}` in record value type, expected `{}`",
+                            rbs_node_type_name(hash_node->value), "RecordFieldtype");
+            }
+
+            continue;
+        }
+
+        rbs_types_record_fieldtype_t *valueNode = (rbs_types_record_fieldtype_t *)hash_node->value;
+        auto innerType = TypeTranslator::toRBI(ctx, typeParams, valueNode->type, loc);
+        valuesStore.emplace_back(std::move(innerType));
+    }
+
+    return ast::MK::Hash(loc, std::move(keysStore), std::move(valuesStore));
+}
+
+ast::ExpressionPtr variableType(core::MutableContext ctx, rbs_types_variable_t *node, core::LocOffsets loc) {
+    rbs_ast_symbol_t *symbol = (rbs_ast_symbol_t *)node->name;
+    rbs_constant_t *constant = rbs_constant_pool_id_to_constant(fake_constant_pool, symbol->constant_id);
+    std::string_view string(constant->start, constant->length);
+    auto name = ctx.state.enterNameUTF8(string);
+    return ast::MK::Send1(loc, ast::MK::T(loc), core::Names::typeParameter(), loc, ast::MK::Symbol(loc, name));
+}
+
+} // namespace
+
+core::LocOffsets TypeTranslator::nodeLoc(core::LocOffsets offset, rbs_node_t *node) {
+    auto loc = node->location;
+    return core::LocOffsets{
+        offset.beginPos() + loc->rg.start.char_pos + 2,
+        offset.beginPos() + loc->rg.end.char_pos + 2,
+    };
+}
+
+ast::ExpressionPtr TypeTranslator::toRBI(core::MutableContext ctx,
+                                         std::vector<std::pair<core::LocOffsets, core::NameRef>> typeParams,
+                                         rbs_node_t *node, core::LocOffsets docLoc) {
+    switch (node->type) {
+        case RBS_TYPES_ALIAS: {
+            auto loc = TypeTranslator::nodeLoc(docLoc, node);
+            if (auto e = ctx.beginError(loc, core::errors::Rewriter::RBSUnsupported)) {
+                e.setHeader("RBS aliases are not supported yet");
+            }
+            return ast::MK::Untyped(docLoc);
+        }
+        case RBS_TYPES_BASES_ANY:
+            return ast::MK::Untyped(docLoc);
+        case RBS_TYPES_BASES_BOOL:
+            return ast::MK::T_Boolean(docLoc);
+        case RBS_TYPES_BASES_BOTTOM:
+            return ast::MK::NoReturn(docLoc);
+        case RBS_TYPES_BASES_CLASS: {
+            auto loc = TypeTranslator::nodeLoc(docLoc, node);
+            if (auto e = ctx.beginError(loc, core::errors::Rewriter::RBSUnsupported)) {
+                e.setHeader("RBS type `{}` is not supported yet", "class");
+            }
+            return ast::MK::Untyped(docLoc);
+        }
```

Comment:
From reading the docs, I believe that for RBS types in `class` bodies (`class A; end`), it's always possible to translate this to `T.class_of(<name of enclosing class>)`.

For modules, Sorbet does not yet have an analogue.